src/System.Linq/src/System/Linq/Partition.cs

```diff
@@ -389,5 +389,245 @@ public int GetCount(bool onlyIfCheap)
+                    }
+                }
+
+                found = false;
+                return default(TSource);
+            }
+
+            public TSource TryGetFirst(out bool found)
+            {
+                using (IEnumerator<TSource> en = _source.GetEnumerator())
+                {
+                    if (SkipBeforeFirst(en) && en.MoveNext())
+                    {
+                        found = true;
+                        return en.Current;
+                    }
+                }
+
+                found = false;
+                return default(TSource);
+            }
+
+            public TSource TryGetLast(out bool found)
+            {
+                using (IEnumerator<TSource> en = _source.GetEnumerator())
+                {
+                    if (SkipBeforeFirst(en))
+                    {
+                        return TryGetLastTaken(en, out found);
+                    }
+                }
+
+                found = false;
+                return default(TSource);
+            }
+
+            public TSource[] ToArray()
+            {
+                using (IEnumerator<TSource> en = _source.GetEnumerator())
+                {
+                    if (SkipBeforeFirst(en) && en.MoveNext())
+                    {
+                        var builder = new LargeArrayBuilder<TSource>(initialize: true);
+                        int toTake = _maxIndexInclusive - _minIndexInclusive + 1;
+
+                        do
+                        {
+                            toTake--;
+                            builder.Add(en.Current);
+                        }
+                        while (toTake > 0 && en.MoveNext());
+
+                        return builder.ToArray();
+                    }
+                }
+
+                return Array.Empty<TSource>();
+            }
+
+            public List<TSource> ToList()
+            {
+                var list = new List<TSource>();
+
+                using (IEnumerator<TSource> en = _source.GetEnumerator())
+                {
+                    if (SkipBeforeFirst(en) && en.MoveNext())
+                    {
+                        int toTake = _maxIndexInclusive - _minIndexInclusive + 1;
+
+                        do
+                        {
+                            toTake--;
+                            list.Add(en.Current);
+                        }
+                        while (toTake > 0 && en.MoveNext());
+                    }
+                }
+
+                return list;
+            }
+
+            private bool SkipBeforeFirst(IEnumerator<TSource> en)
+            {
+                Debug.Assert(en != null);
+
+                int toSkip = _minIndexInclusive;
+                while (toSkip > 0 && en.MoveNext())
+                {
+                    toSkip--;
+                }
+
+                // Did we finish skipping before we reached the end of the sequence?
+                return toSkip == 0;
+            }
```

Comment:
@stephentoub Good idea. In fact, I renamed the method to be `SkipBefore` and had `TryGetElementAt` do `SkipBefore(_minIndexInclusive + index, en)`. Should be slightly more efficient.